frontend/packages/kubevirt-plugin/src/k8s/enhancedK8sMethods/__tests__/enhancedK8sMethods.spec.ts

```diff
@@ -0,0 +1,205 @@
+    ]);
+
+    expect(methodsUnchecked.history).toHaveLength(3);
+  });
+
+  it('shows actualState', async () => {
+    const { methods, methodsUnchecked } = spyEnhancedK8sMethods();
+
+    await methods.k8sCreate(VirtualMachineModel, testVM);
+    await methods.k8sPatch(VirtualMachineModel, testVM, []);
+    await methods.k8sPatch(VirtualMachineModel, testVM, []);
+    await methodsUnchecked.k8sGet(VirtualMachineModel, testVM);
+    await methods.k8sKill(VirtualMachineModel, testVM);
+    await methods.k8sCreate(VirtualMachineModel, otherTestVM);
+    await methods.k8sPatch(VirtualMachineModel, otherTestVM, []);
+    await methodsUnchecked.k8sGet(VirtualMachineModel, otherTestVM);
+    await methods.k8sCreate(VirtualMachineModel, testVM);
+    await methodsUnchecked.k8sGet(VirtualMachineModel, testVM);
+    await methods.k8sPatch(VirtualMachineModel, testVM, [testPatch]);
+    const actualState = methods.getActualState();
+
+    expect(actualState).toHaveLength(2);
+    expect(actualState[0]).toEqual(otherTestVM);
+    expect(actualState[1]).toEqual({ ...testVM, ...testPatch });
+  });
+
+  it('disables recording history', async () => {
+    const { methods, methodsUnchecked } = spyEnhancedK8sMethods();
+
+    await methods.k8sCreate(VirtualMachineModel, testVM, null, disableHistoryOpts);
+    await methods.k8sPatch(VirtualMachineModel, testVM, [testPatch], disableHistoryOpts);
+    await methods.k8sKill(VirtualMachineModel, testVM, null, null, disableHistoryOpts);
+
+    expect(methods.getHistory()).toHaveLength(0);
+    expect(methodsUnchecked.history).toHaveLength(0);
+
+    await methods.k8sCreate(VirtualMachineModel, testVM);
+    expect(methods.getHistory()).toHaveLength(1);
+  });
+
+  it('rollback', async () => {
+    const { methods } = spyEnhancedK8sMethods();
+
+    await methods.k8sCreate(VirtualMachineModel, testVM);
+    await methods.k8sPatch(VirtualMachineModel, testVM, []);
+    await methods.k8sCreate(VirtualMachineModel, otherTestVM);
+
+    const results = await methods.rollback();
+
+    expect(results).toHaveLength(2);
+    expect(methods.getActualState()).toHaveLength(0);
+
+    expectHistory(methods.getHistory(), [
+      new HistoryItem(HistoryType.CREATE, testVM),
+      new HistoryItem(HistoryType.PATCH, testVM),
+      new HistoryItem(HistoryType.CREATE, otherTestVM),
+      // reverse order of deletion
+      new HistoryItem(HistoryType.DELETE, otherTestVM),
+      new HistoryItem(HistoryType.DELETE, testVM),
+    ]);
+  });
+
+  it('rollback fails', async () => {
+    const { methods, methodsUnchecked } = spyEnhancedK8sMethods();
+    methodsUnchecked.k8sKill = async () => Promise.reject(new Error('delete failed'));
+
+    await methods.k8sCreate(VirtualMachineModel, testVM);
+    await methods.k8sPatch(VirtualMachineModel, testVM, []);
+    await methods.k8sCreate(VirtualMachineModel, otherTestVM);
+
+    expect.assertions(4);
+    let rollbackError;
+    try {
+      await methods.rollback();
```

Comment:
We could use the [`toThrowError`](https://jestjs.io/docs/en/expect#tothrowerror) matcher here.

```ts
expect(methods.rollback).toThrowError('rollback');
// alternatively perform equality check against provided object
expect(methods.rollback).toThrowError(new Error('rollback'));
```

Reply:
This does not work very well as the error must implement Constructable